### PR TITLE
Implements filtering tests for mag/min/mipmapFilters with additional float32-filterable formats.

### PR DESCRIPTION
**Pull request by lokokung**
Issue: #2072

<hr>

**Requirements for PR author:**

- [x] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [x] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [x] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

### CODE REVIEW

File: src/webgpu/api/operation/sampling/filter_mode.spec.ts [hunk 1]
```diff
@@ -1,14 +1,629 @@
...
+      struct VertexOut {
+        @builtin(position) pos: vec4f,
+        @location(0) uv: vec2f,
+      };
+
+      @vertex
+      fn vs_main(@builtin(vertex_index) vi : u32,
+                 @builtin(instance_index) ii: u32) -> VertexOut {
+        const grid = vec2f(4., 4.);
+        const size = 1. / 2.;
```
Comment (austinEng) [inline src/webgpu/api/operation/sampling/filter_mode.spec.ts:365]: would help to describe what "space" these are in to help make sense of the transformations

File: src/webgpu/api/operation/sampling/filter_mode.spec.ts [hunk 2]
```diff
@@ -1,14 +1,629 @@
...
+
+        const constantPlaneOffset = vec2f(3. / 8., 3. / 8.);
+        const perPixelOffset = -1. / 4.;
```
Comment (austinEng) [inline src/webgpu/api/operation/sampling/filter_mode.spec.ts:368]: it's not obvious _why_ these values are the way they are without seeing a picture what what we're trying to do - importantly how you're placing the smaller sub-pixel quad inside the pixel.

Maybe you can have a small ascii art picture to depict the goal

File: src/webgpu/util/texture/texel_data.ts [hunk 3]
```diff
@@ -926,3 +926,21 @@ export function getComponentReadbackTraits(dataType: ComponentDataType) {
       unreachable();
   }
 }
+
+/**
+ * Get generator of all the coordinates in a subrect.
+ * @param subrectOrigin - Subrect origin
+ * @param subrectSize - Subrect size
+ */
+export function* fullSubrectCoordinates(
```
[the remark this answers is not in the document]
Comment (lokokung) [reply src/webgpu/util/texture/texel_data.ts]: Done.

File: src/webgpu/api/operation/sampling/filter_mode.spec.ts [hunk 4]
```diff
@@ -1,14 +1,629 @@
...
+const kExpectedUVRepeatLinear = [
+  [ 0.375, 0.375, 0.625, 0.625 ],
+  [ 0.375, 0.375, 0.625, 0.625 ],
+  [ 0.625, 0.625, 0.375, 0.375 ],
+  [ 0.625, 0.625, 0.375, 0.375 ],
+];
+/* eslint-enable prettier/prettier */
+
+// Given the 2x2 checkerboard texture we are using for the sample texture, address modes of
+// 'clamped-to-edge' and 'mirror-repeat' are actually identical, so we can just do a preprocess
```
[the remark this answers is not in the document]
Comment (lokokung) [reply src/webgpu/api/operation/sampling/filter_mode.spec.ts]: Took a while... but done now!

File: src/webgpu/api/operation/sampling/filter_mode.spec.ts (hunk 1 above)
Comment (lokokung) [reply src/webgpu/api/operation/sampling/filter_mode.spec.ts]: Added some more comments and some ascii art to help.

File: src/webgpu/api/operation/sampling/filter_mode.spec.ts (hunk 2 above)
Comment (lokokung) [reply src/webgpu/api/operation/sampling/filter_mode.spec.ts]: Done

**Comment by lokokung** [comment]
Ah, sorry it too so long to fix! Pretty much needed to rewrite all the min/magFilter tests in order to differentiate between `mirror-repeat` and `clamp-to-edge`. We can chat offline more if there's anything in the tests that is too confusing!

File: src/webgpu/api/operation/sampling/filter_mode.spec.ts [hunk 5]
```diff
@@ -1,14 +1,629 @@
...
+      vertex: {
+        module,
+        entryPoint: 'vs_main',
+      },
+      fragment: {
+        module,
+        entryPoint: 'fs_main',
+        targets: [{ format }],
+      },
+      primitive: { topology: 'triangle-list' },
```
[the remark this answers is not in the document]
Comment (lokokung) [reply src/webgpu/api/operation/sampling/filter_mode.spec.ts]: Done

**Review by austinEng — APPROVED** [review]
LGTM! thank you for the incredible comments and diagrams explaining the tests!

**Pull request merged**